File: src/intellijeval/toolwindow/NewElementPopupAction.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package intellijeval.toolwindow;

import com.intellij.ide.IdeBundle;
import com.intellij.openapi.actionSystem.*;
import com.intellij.openapi.fileChooser.actions.NewFolderAction;
import com.intellij.openapi.project.DumbAware;
import com.intellij.openapi.ui.popup.JBPopupFactory;
import com.intellij.openapi.ui.popup.ListPopup;
import com.intellij.openapi.util.IconLoader;
import intellijeval.Util;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.util.ArrayList;
import java.util.Arrays;

/**
 * User: dima
 * Date: 13/08/2012
 */
class NewElementPopupAction extends AnAction implements DumbAware, PopupAction {
	public static final Icon Folder = IconLoader.getIcon("/nodes/folder.png"); // 16x16

	@Override public void actionPerformed(final AnActionEvent event) {
		showPopup(event.getDataContext());
	}

	private void showPopup(DataContext context) {
		createPopup(context).showInBestPositionFor(context);
	}

	private ListPopup createPopup(DataContext dataContext) {
		return JBPopupFactory.getInstance().createActionGroupPopup(
				IdeBundle.message("title.popup.new.element"),
				createActionGroup(),
				dataContext,
				false,
				true,
				false,
				null,
				-1,
				LangDataKeys.PRESELECT_NEW_ACTION_CONDITION.getData(dataContext)
		);
	}

	private ActionGroup createActionGroup() {
		return new ActionGroup() {
			@NotNull @Override
			public AnAction[] getChildren(AnActionEvent e) {
				ArrayList<AnAction> actions = new ArrayList<AnAction>();
				actions.addAll(Arrays.asList(
						new NewFileAction("Groovy Script", Util.GROOVY_FILE_TYPE_ICON),
						new NewFolderAction("Directory", "Directory", Folder),
						new Separator(),
						new PluginToolWindowManager.AddNewPluginAction(),
						new PluginToolWindowManager.AddPluginFromDiskAction()
				));
				if (PluginToolWindowManager.addFromGitHubAction != null) {
					actions.add(PluginToolWindowManager.addFromGitHubAction);
				}
				return actions.toArray(new AnAction[actions.size()]);
			}
		};
	}

	@Override public void update(AnActionEvent e) {
		Presentation presentation = e.getPresentation();
		presentation.setEnabled(true);
	}
}

File: src/main/liveplugin/implementation/toolwindow/popup/NewElementPopupAction.kt
package liveplugin.implementation.toolwindow.popup

import com.intellij.ide.IdeBundle
import com.intellij.openapi.actionSystem.*
import com.intellij.openapi.fileChooser.actions.NewFolderAction
import com.intellij.openapi.project.DumbAware
import com.intellij.openapi.ui.popup.JBPopupFactory
import com.intellij.openapi.ui.popup.ListPopup
import liveplugin.implementation.LivePluginPaths.groovyExamplesPath
import liveplugin.implementation.LivePluginPaths.kotlinExamplesPath
import liveplugin.implementation.common.Icons.newFolderIcon
import liveplugin.implementation.common.IdeUtil.groovyFileType
import liveplugin.implementation.common.IdeUtil.kotlinFileType
import liveplugin.implementation.common.IdeUtil.textFileType
import liveplugin.implementation.common.IdeUtil.xmlFileType
import liveplugin.implementation.pluginrunner.groovy.GroovyPluginRunner
import liveplugin.implementation.pluginrunner.kotlin.KotlinPluginRunner
import liveplugin.implementation.toolwindow.addplugin.*
import liveplugin.implementation.toolwindow.util.readSampleScriptFile

class NewElementPopupAction: AnAction(), DumbAware, PopupAction {
    override fun actionPerformed(event: AnActionEvent) {
        createPopup(event.dataContext).showInBestPositionFor(event.dataContext)
    }

    private fun createPopup(dataContext: DataContext): ListPopup =
        JBPopupFactory.getInstance().createActionGroupPopup(
            IdeBundle.message("title.popup.new.element"),
            livePluginNewElementPopup,
            dataContext, false, true, false, null, -1,
            LangDataKeys.PRESELECT_NEW_ACTION_CONDITION.getData(dataContext)
        )

    override fun update(event: AnActionEvent) {
        event.presentation.isEnabled = true
    }

    companion object {
        val livePluginNewElementPopup by lazy {
            DefaultActionGroup(
                { "New" },
                listOf(
                    NewKotlinFileAction(),
                    NewGroovyFileAction(),
                    NewTextFileAction(),
                    NewDirectoryAction(),
                    NewKotlinMainScript(),
                    NewGroovyMainScript(),
                    NewGroovyTestScript(),
                    NewPluginXmlScript(),
                    Separator.getInstance(),
                    AddNewKotlinPluginAction(),
                    AddNewGroovyPluginAction(),
                    AddPluginFromGistDelegateAction(),
                    AddPluginFromGitHubDelegateAction(),
                    AddKotlinExamplesActionGroup(),
                    AddGroovyExamplesActionGroup(),
                )
            ).also { it.isPopup = true }
        }
    }

    private class NewGroovyFileAction: NewFileAction("Groovy File", groovyFileType)

    private class NewKotlinFileAction: NewFileAction("Kotlin File", kotlinFileType)

    private class NewTextFileAction: NewFileAction("Text File", textFileType)

    private class NewDirectoryAction: NewFolderAction("Directory", "", newFolderIcon)

    private class NewGroovyMainScript: NewFileFromTemplateAction(
        GroovyPluginRunner.mainScript,
        GroovyPluginRunner.mainScript,
        readSampleScriptFile("$groovyExamplesPath/default-plugin.groovy"),
        groovyFileType,
    )

    private class NewKotlinMainScript: NewFileFromTemplateAction(
        KotlinPluginRunner.mainScript,
        KotlinPluginRunner.mainScript,
        readSampleScriptFile("$kotlinExamplesPath/default-plugin.kts"),
        kotlinFileType
    )

    private class NewGroovyTestScript: NewFileFromTemplateAction(
        GroovyPluginRunner.testScript,
        GroovyPluginRunner.testScript,
        readSampleScriptFile("$groovyExamplesPath/default-plugin-test.groovy"),
        groovyFileType
    )
}

class NewPluginXmlScript(
    fileContent: String = readSampleScriptFile("$kotlinExamplesPath/plugin.xml")
): NewFileFromTemplateAction(
    "plugin.xml",
    "plugin.xml",
    fileContent,
    xmlFileType
)
